### DISCUSSION TITLE
make a web app

### DISCUSSION THREAD

**Discussion by PencilTorch** [reo7sp/tgbot-cpp#302] [Q&A] [answered]
I can't figure out how to create a web app with sending and receiving data. Is there an example?

**Comment by reo7sp** [comment, marked as answer]
tgbot-cpp implements only the bot backend; the Web App itself is a separate HTTPS page using the Telegram Web App JavaScript API.

For the simplest `sendData` flow, attach a `WebAppInfo` to a reply-keyboard button:

```cpp
auto app = std::make_shared<TgBot::WebAppInfo>();
app->url = "https://example.com/app";

auto button = std::make_shared<TgBot::KeyboardButton>();
button->text = "Open app";
button->webApp = app;
```

In JavaScript call `Telegram.WebApp.sendData(...)`. The bot receives a normal message with `message->webAppData`; check it in `onAnyMessage` and read `message->webAppData->data`.

An `InlineKeyboardButton::webApp` uses the inline Web App query flow instead: send the query ID from the Web App to your backend and respond with `bot.getApi().answerWebAppQuery(...)`. Always validate Telegram `initData` on the backend; do not trust `initDataUnsafe` by itself.
